azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	if vulnerabilityAssessment.ID != nil {
+		d.SetId(*vulnerabilityAssessment.ID)
+	}
+
+	// Take some time for resource to be visible after creating. Hence the retry logic
+	readError := resourceServerVulnerabilityAssessmentRead(d, meta)
+	for retry := 1; d.Id() == "" && retry < 11; retry++ {
+		time.Sleep(time.Duration(retry) * time.Second)
+		readError = resourceServerVulnerabilityAssessmentRead(d, meta)
+	}
+	return readError
+}
+
+func resourceServerVulnerabilityAssessmentRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).SecurityCenter.ServerVulnerabilityAssessmentClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	vmId := d.Get("virtual_machine_id").(string)
+	var idToParse string
+	if vmId != "" {
+		idToParse = vmId
+	} else {
+		idToParse = d.Id()
+	}
+	resourceGroup, provider, resourceType, name, id, err := parseVirtualMacineId(idToParse)
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, provider, resourceType, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("Server Vulnerability Assessment was not found for %q/%q: %+v", resourceGroup, name, err)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving Server Vulnerability Assessment status for %q/%q: %+v", resourceGroup, name, err)
+	}
+
+	d.SetId(*resp.ID)
+	if vmId == "" {
+		// Need to set virtual_machine_id when we do a import
+		_ = d.Set("virtual_machine_id", "/subscriptions/"+id.SubscriptionID+
+			"/resourceGroups/"+id.ResourceGroup+
+			"/providers/"+provider+"/"+resourceType+"/"+id.Path[resourceType])
+	}
+
+	return nil
+}
+
+func resourceServerVulnerabilityAssessmentDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).SecurityCenter.ServerVulnerabilityAssessmentClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	resourceGroup, provider, resourceType, name, _, err := parseVirtualMacineId(d.Get("virtual_machine_id").(string))
+	if err != nil {
+		return err
+	}
+
+	// Can not delete if still in provisioning state. Wait for it to complete.
+	for retry := 1; retry <= 20; retry++ {
+		response, err := client.Get(ctx, resourceGroup, provider, resourceType, name)
+		if err != nil || response.ProvisioningState != security.ProvisioningState1Provisioning {
+			break
+		}
+		time.Sleep(time.Duration(retry) * time.Second)
+	}
+
+	response, err := client.Delete(ctx, resourceGroup, provider, resourceType, name)
+	if err != nil {
+		if utils.ResponseWasStatusCode(response, 202) {
+			return nil
+		} else {
+			return fmt.Errorf("removing Server Vulnerability Assessment for %q/%q: %+v\n", resourceGroup, name, err)
+		}
+	} else {
+		return nil
+	}
+}
+
+func parseVirtualMacineId(stringId string) (string, string, string, string, *azure.ResourceID, error) {
```

Comment:
Typo here `parseVirtualMachineId` however.... There already exists a parser for virtualMachines `azurerm/internal/services/compute/parse/virtual_machine.go:41` you should reuse that. This however does not support Arc.

If you want to support Arc then I would look into using the generator for resource id parsers/validators `azurerm/internal/tools/generator-resource-id/README.md`. Then I would split the input parameters into conflicting `virtual_machine_id` and `hybrid_machine_id` allowing you to use the generated validators for each id type. Although _personally_ I would probably drop the Arc support from this PR to keep it small, replacing this parser with the existing parser. Then if Arc support is desired, implement that as a new PR on top of these changes.

Reply:
@neil-yechenwei is Arc support something you need, or can we wait with that?

Reply:
@sharebear I would appreciate if you took a look at my changes in commit 2ffd690 and let me know if you think they look good.

Reply:
It's probably best to wait and see what the maintainers say, and I'm still very new to golang and writing providers. That said, I have a gut feeling that the implementation feels a little complicated compared to other resources I've looked at.... what if read looked like the following

```
targetId = ""
virtualMachineId, err := computeParse.VirtualMachineID(d.Id())
if virtualMachineId != nil {
	targetId = virtualMachineId.ID()
	d.Set("virtual_machine_id", targetId)
}

if targetId == "" {
	hybridMachineId, err := computeParse.HybridMachineID(d.Id())
	if hybridMachineId != nil {
		targetId = hybridMachineId.ID()
		d.Set("hybrid_machine_id", targetId)
	}
}

// TODO: Return err if targetId still emtpy

// do client.Get
```

I think create would be even easier to simplify, then you could drop the whole TargetId struct.... disclaimer, I'm shooting from the hip here, no idea if this would work in practice.

Reply:
Agree, my problem is that the API needs a couple specific parts of the id and not the entire id. Will keep like this and see what the maintainers say as you suggested.